### ISSUE TITLE
Microsoft.Extensions.Logging ILogger scopes are lost in async stream after `yield return`

### ISSUE THREAD

**Issue by CXuesong** [dotnet/runtime#47951]
Opened via https://github.com/dotnet/runtime/issues/47802#issuecomment-773450789. While the previous issue focuses on the decision of whether `AsyncLocal`s (and execution context) should flow acress `yield return` in async streams, I hope the current one can be triaged to MEL owners, in order to decide on whether the following scoped logging scenario should be supported.

### Description
Consider the following use case on .NET 5.0 [[.NET Fiddle](https://dotnetfiddle.net/DauJqS)]
```c#
using System;
using System.Collections.Generic;
using Microsoft.Extensions.Logging;

async IAsyncEnumerable<int> IntSequence(ILogger logger)
{
    using var scope = logger.BeginScope("My scope");
    logger.LogWarning("Log 1");
    yield return 10;
    logger.LogWarning("Log 2");
    yield return 20;
    logger.LogWarning("Log 3");
    yield return 30;
}

using var lf = LoggerFactory.Create(b => b.AddSimpleConsole(o => o.IncludeScopes = true));
await foreach (var i in IntSequence(lf.CreateLogger("Root")))
{
    Console.WriteLine(i);
}
```

My expectation is
```
10
20
30
warn: Root[0]
      => My scope
      Log 1
warn: Root[0]
      => My scope
      Log 2
warn: Root[0]
      => My scope
      Log 3
```
Actual output is
```
10
20
30
warn: Root[0]
      => My scope
      Log 1
warn: Root[0]
      Log 2
warn: Root[0]
      Log 3
```
Scopes are lost for Log 2 and Log 3.

### Configuration
.NET 5 / .NET 6.0.100-preview.1.21103.9

### Regression?
N/A

### Other information

The root cause of this issue is #47802, and `LoggerFactoryScopeProvider` uses `AsyncLocal` to keep logging scope information.
https://github.com/dotnet/runtime/blob/f93fee40a84e04c3a28bc4b41b182ca38424c69a/src/libraries/Microsoft.Extensions.Logging/src/LoggerFactoryScopeProvider.cs#L14-L21

**Comment by huoyaoyuan** [comment]
The context here are actually depending on the `ConfigureAwait` on `await foreach`. It it uses `ConfigureAwait(true)` AND there is valid context, the continuation after yield return will run on the same context(thread).

**Comment by CXuesong** [comment]
@huoyaoyuan I think `ConfigureAwait` controls whether to execute continuation on the same `SynchronizationContext` rather than `ExecutionContext`. And `AsyncLocal` (or `ThreadLocal`) depends on `ExecutionContext` to flow.

If you expand the `foreach` syntax, you will see exactly the same situation
```
using var lf = LoggerFactory.Create(b => b.AddSimpleConsole(o => o.IncludeScopes = true));
await using (var it = IntSequence(lf.CreateLogger("Root")).GetAsyncEnumerator())
{
	while (await it.MoveNextAsync().ConfigureAwait(true)) {
    	Console.WriteLine(it.Current);
	}
}
```

**Comment by stephentoub** [comment]
I assume this behaves as you'd expect if you added `await Task.Yield();` after you create the scope?

**Comment by CXuesong** [comment]
> I assume this behaves as you'd expect if you added `await Task.Yield();` after you create the scope?

@stephentoub Yes. `await`s does not affect logging scopes.

Actually `AsyncStateMachineBox<T>` is keeping ExecutionContext and restoring it in `MoveNext` call.

**Comment by stephentoub** [comment]
Right,  my point was just that a workaround is to make sure you have an await like that after anything that changes EC.

**Comment by CXuesong** [comment]
> Right, my point was just that a workaround is to make sure you have an await like that after anything that changes EC.
@stephentoub Oh… Did you mean something like this?
```c#
async IAsyncEnumerable<int> IntSequence(ILogger logger)
{
    using var scope = logger.BeginScope("My scope");
    logger.LogWarning("Log 1");
    yield return 10;
    await Task.Yield()        // <---- yield here
    logger.LogWarning("Log 2");
    yield return 20;
    logger.LogWarning("Log 3");
    yield return 30;
}
```
Then this doesn't work… I meant `await` captures the EC before leaving the function, and restores it when re-entering. Since we have already lost the EC after `yield return`, the EC captured before `await` is the default EC without `AsyncLocal` values.

**Comment by stephentoub** [comment]
> Did you mean something like this?

No, as the first line after the using.

**Comment by CXuesong** [comment]
> > Did you mean something like this?
> 
> No, as the first line after the using.

Oh. Then the logging scopes are still lost since "Log 2", as if `Yield` never exists. I've corrected my answer above.

I assume you meant this:
```c#
async IAsyncEnumerable<int> IntSequence(ILogger logger)
{
    using var scope = logger.BeginScope("My scope");
    await Task.Yield();        // <-- yield here
    logger.LogWarning("Log 1");
    yield return 10;
    logger.LogWarning("Log 2");
    yield return 20;
    logger.LogWarning("Log 3");
    yield return 30;
}
```

**Comment by springy76** [comment]
When does an abstract context start and when does it end? https://learn.microsoft.com/en-us/dotnet/api/system.threading.asynclocal-1?view=net-7.0 doesn't give a hint you have to use `ConfigureAwait(true)`.

How does the correct usage of `ILogger.BeginScope()` inside an async method look like?

While most of the time everything looks right, I just noticed (in SEQ, which aggregates structured logging) that sometimes totally unrelated (structured) scope keys are added to log entries which have no code path which could add them - so the only source can be a reused thread (out of the thread pool of course) which has remaining values from "foreign" AsyncLocals.

**Comment by ja-banika** [comment]
refactor public async Task<IActionResult> Post([FromQuery] string validationToken = null)
    {
      // handle validation
      if (!string.IsNullOrEmpty(validationToken))
      {
        _logger.LogInformation($"LifecycleNotificationsController Received Token: '{validationToken}'");
        return Ok(validationToken);
      }

      var content = await Request.Body.ReadBody();

      _logger.LogInformation("Message body received from lifecycle notification webhook: {@MessageBody}", content);

      var notifications = JsonConvert.DeserializeObject<LifecycleEventNotifications>(content);

      if (notifications?.Items != null)
      {
        foreach (var notification in notifications.Items)
        {
          var dictionary = new Dictionary<string, object>();
          var msalAccountActivity = new MsalAccountActivity();
          if (notification.ClientState != null)
          {
            msalAccountActivity = await _msalAccountActivityRepository.GetMsalAccountActivityByClientUserSecret(
              Guid.Parse(notification.ClientState));

            if (msalAccountActivity != null)
            {
              dictionary.Add(LoggingConstants.AccountId, msalAccountActivity?.UserAccountId);
              dictionary.Add(LoggingConstants.UserId, msalAccountActivity?.UserId);
              dictionary.Add(LoggingConstants.Email, msalAccountActivity?.UserPrincipalName);
              dictionary.Add(LoggingConstants.Instance, msalAccountActivity?.Instance);
            }
          }

          using (_logger.BeginScope(dictionary))
          {
            _logger.LogInformation("{@Notification}", notification);
            if (notification.ShouldHandleEvent())
            {
              await _eventBridgeService.EmitMessage(new LifecycleNotificationRequest(notification.ToLifecycleEventMessage(), msalAccountActivity.ToJobAdderUser()));
            }
          }
        }
      }
      else
      {
        _logger.LogInformation("No notification items found for notifications: {@Notifications}", notifications);
      }

      return Ok();
    }

**Comment by DavidThielen** [comment]
Hi;

One suggestion/request. I just spent about an hour trying to figure out why my BeginScope scope was being discarded. I was trying to have it persist for the lifetime of the ILogger.

Not a problem that it is designed to have a very short lifespan. It's like DbContext. But please state that in the documentation. Once I realized that I came up with what is actually a better implementation for what I need. So no problem that it's ethereal. Just document that please.

thanks - dave
